collector/processor/faasprocessor/processor.go

```diff
@@ -0,0 +1,171 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package faasprocessor // import "github.com/open-telemetry/opentelemetry-lambda/collector/processor/faasprocessor"
+
+import (
+	"context"
+
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/consumer"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/ptrace"
+	"go.opentelemetry.io/collector/processor"
+	"go.opentelemetry.io/collector/processor/processorhelper"
+	semconv "go.opentelemetry.io/collector/semconv/v1.22.0"
+	"go.uber.org/zap"
+)
+
+const (
+	telemetryAPIScope = "github.com/open-telemetry/opentelemetry-lambda/collector/receiver/telemetryapi"
+)
+
+type faasProcessor struct {
+	telemetryAPIRuntimeSpans map[string]cachedSpan
+	telemetryAPIInitSpans    map[string]cachedSpan
+	invocationSpans          map[string]cachedSpan
+	logger                   *zap.Logger
+	nextConsumer             consumer.Traces
+}
+
+func newFaasProcessor(next consumer.Traces, set processor.CreateSettings) (*faasProcessor, error) {
+	return &faasProcessor{
+		telemetryAPIRuntimeSpans: make(map[string]cachedSpan),
+		telemetryAPIInitSpans:    make(map[string]cachedSpan),
+		invocationSpans:          make(map[string]cachedSpan),
+		nextConsumer:             next,
+		logger:                   set.Logger,
+	}, nil
+}
+
+/* -------------------------------------- TRACES INTERFACE ------------------------------------- */
+
+func (*faasProcessor) Capabilities() consumer.Capabilities {
+	return consumer.Capabilities{MutatesData: true}
+}
+
+func (p *faasProcessor) ConsumeTraces(ctx context.Context, td ptrace.Traces) error {
```

Comment:
Simplifying this method will help with reasoning about this code and testing paths. For example, these high-level operations make sense to me:
- categorizeSpan (caches with runtimeSpans, initSpans, or invocationSpans)
- associateSpans (adds traceid and parentid)
- forwardRemainingSpans